### PR TITLE
Gradient/Jacobians with pytrees

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**
No matter what is the structure of the return, when taking derivatives we also return the flatten derivatives.

**Description of the Change:**
Add support for any return when taking the derivatives and keep the structure. (Pytree support)

```
dev = qml.device("lightning.qubit", wires=1)

@qml.qnode(dev)
def circuit(phi, psi):
    qml.RY(phi, wires=0)
    qml.RX(psi, wires=0)
    return [{"expval0": qml.expval(qml.PauliZ(0))}, qml.expval(qml.PauliZ(0))]

psi = 0.1
```
```
print(qjit(jacobian(circuit, argnum=[0, 1]))(psi, phi))
[{'expval0': (array(-0.0978434), array(-0.19767681))}, (array(-0.0978434), array(-0.19767681))]
```

### CODE REVIEW

**Comment by rmoyard** [comment]
[sc-55112]

**Review by erick-xanadu — APPROVED** [review]
:partying_face:

File: frontend/catalyst/pennylane_extensions.py [hunk 1]
```diff
@@ -522,6 +523,27 @@ def _check_grad_params(
     return GradParams(method, scalar_out, h, argnum)
 
 
+def _unflatten_derivatives(results, out_tree, argnum, num_results):
+    """Unflatten the flat list of derivatives results given the out tree."""
+    num_trainable_params = len(argnum) if isinstance(argnum, list) else 1
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:528]: Checking for a list is unnecessarily restrictive, maybe better to check for any iterable?

File: frontend/test/pytest/test_gradient.py
Comment (dime10) [inline frontend/test/pytest/test_gradient.py:1]: We should include tests for Pytree arguments as well, unless we already have those?

File: frontend/catalyst/pennylane_extensions.py [hunk 2]
```diff
@@ -522,6 +523,27 @@ def _check_grad_params(
...
+        tuple(results[i * num_trainable_params : i * num_trainable_params + num_trainable_params])
+        for i in range(0, num_results)
+    )
+    # In jax,  argnums=[int] wraps single derivatives in a tuple
+    if (
+        (isinstance(argnum, int) or argnum is None)
+        and num_trainable_params == 1
+        and num_results != 1
+    ):
+        results = tuple(r[0] if len(r) == 1 else r for r in results)
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:539]: I find these lines very difficult to follow, maybe there is a way to refactor them or explain what they are doing?

File: frontend/catalyst/pennylane_extensions.py [hunk 3]
```diff
@@ -522,6 +523,27 @@ def _check_grad_params(
...
+    if out_tree.children() != []:
+        results = tree_unflatten(out_tree, results)
+    else:
+        results = results[0]
```
Comment (dime10) [inline frontend/catalyst/pennylane_extensions.py:543]: Is this sort of special case really required? I would have hoped that unflattening the tree would take care of it.

Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:543]: Yes me too, but in some cases there is an error internal to Jax.

File: frontend/catalyst/pennylane_extensions.py (hunk 2 above)
Comment (rmoyard) [reply frontend/catalyst/pennylane_extensions.py:539]: This is doing:

```
flat list of derivatives -> structure of the jacobian (tuple of tuple if multiple return and multiple parameters)
```

**Pull request merged**